src/components/TodoList/TodoList.tsx

```diff
@@ -1,44 +1,168 @@
-import React from 'react';
+import React, { useCallback, useEffect, useState } from 'react';
+import classNames from 'classnames';
+import { getTodo } from '../../api/api';
+import { Todo } from '../../react-app-env';
+
 import './TodoList.scss';
 
-export const TodoList: React.FC = () => (
-  <div className="TodoList">
-    <h2>Todos:</h2>
-
-    <div className="TodoList__list-container">
-      <ul className="TodoList__list">
-        <li className="TodoList__item TodoList__item--unchecked">
-          <label>
-            <input type="checkbox" readOnly />
-            <p>delectus aut autem</p>
-          </label>
-
-          <button
-            className="
-              TodoList__user-button
-              TodoList__user-button--selected
-              button
-            "
-            type="button"
-          >
-            User&nbsp;#1
-          </button>
-        </li>
-
-        <li className="TodoList__item TodoList__item--checked">
-          <label>
-            <input type="checkbox" checked readOnly />
-            <p>distinctio vitae autem nihil ut molestias quo</p>
-          </label>
-
-          <button
-            className="TodoList__user-button button"
-            type="button"
-          >
-            User&nbsp;#2
-          </button>
-        </li>
-      </ul>
+type Props = {
+  onUserIdChange: (userId: number) => void,
+};
+
+export const TodoList: React.FC<Props> = React.memo(({ onUserIdChange }) => {
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [query, setQuery] = useState('');
+  const [filterByStatus, setFilterByStatus] = useState('all');
+  const [visibleTodos, setVisibleTodos] = useState<Todo[]>([]);
+
+  useEffect(() => {
+    const fetchTodos = async () => {
+      const todosFromServer = await getTodo();
+
+      setTodos(todosFromServer);
+      setVisibleTodos(todosFromServer);
+    };
+
+    try {
+      fetchTodos();
+    } catch (error) {
+      throw new Error('Failed to load todos from server');
+    }
+  }, []);
+
+  const onFilteringByQuery = () => {
+    setVisibleTodos([...visibleTodos].filter(
+      todo => todo.title.toLowerCase().includes(query.toLowerCase()),
+    ));
+  };
+
+  const onFilteringByStatus = (sortBy: string) => {
+    switch (sortBy) {
+      case 'active':
+        setVisibleTodos([...todos].filter(todo => !todo.completed));
+        break;
+
+      case 'completed':
+        setVisibleTodos([...todos].filter(todo => todo.completed));
+        break;
+
+      default:
+        setVisibleTodos([...todos]);
+    }
+  };
```

Comment:
Move this to handleFilterChange

Reply:
This approach causes problems. We try to filter the array by the new state before it is updated. So I combined both filtering methods inside a `useEffect `hook